source/server/lds_api.cc

```diff
@@ -31,63 +31,84 @@ LdsApiImpl::LdsApiImpl(const envoy::api::v2::core::ConfigSource& lds_config,
   init_manager.add(init_target_);
 }
 
-void LdsApiImpl::onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
-                                const std::string& version_info) {
+void LdsApiImpl::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+    const std::string& system_version_info) {
   cm_.adsMux().pause(Config::TypeUrl::get().RouteConfiguration);
   Cleanup rds_resume([this] { cm_.adsMux().resume(Config::TypeUrl::get().RouteConfiguration); });
 
-  std::vector<envoy::api::v2::Listener> listeners;
-  for (const auto& listener_blob : resources) {
-    listeners.push_back(MessageUtil::anyConvert<envoy::api::v2::Listener>(listener_blob));
-    MessageUtil::validate(listeners.back());
-  }
-  std::vector<std::string> exception_msgs;
-  std::unordered_set<std::string> listener_names;
-  for (const auto& listener : listeners) {
-    if (!listener_names.insert(listener.name()).second) {
-      throw EnvoyException(fmt::format("duplicate listener {} found", listener.name()));
+  // We do all listener removals before adding the new listeners. This allows adding a new listener
+  // with the same address as a listener that is to be removed. Do not change the order.
+  for (const auto& removed_listener : removed_resources) {
+    if (listener_manager_.removeListener(removed_listener)) {
```

Comment:
Out of curiosity, do we have disjointness checks in the delta xDS code that looks for overlapping names in added/removed resources? Seems like we would want to reject such an update early.

Reply:
We do not. I will add that separately.

Reply:
#6955